# Activitat 2: Streaming - adquisició i anàlisi de dades

# Introducció

En aquesta activitat, explorarem dues de les tecnologies més populars utilitzades per a l'adquisició i anàlisi de dades en temps real: Apache Kafka i Apache Spark Streaming. A través d'una sèrie d'exercicis, aprendreu els conceptes bàsics d'aquestes tecnologies i com usar-les per processar dades en temps real.

## Estructura

Hem dividit aquesta activitat en tres parts:

1.  **Introducció a Apache Kafka:** En aquesta primera part, aprendreu els conceptes bàsics d'Apache Kafka, 
    els seus principals components i com interactuar amb ells utilitzant la interfície de línia de comandes.
2.  **Ingesta de dades amb Apache Kafka:** En aquesta segona part, aprendreu a utilitzar Apache Kafka per ingerir
    dades des d'un productor i consumir-les des d'un consumidor.
3.  **Processament de dades en temps real amb Apache Spark Streaming:** En aquesta tercera part, aprendreu a usar 
    Apache Spark Streaming per processar dades en temps real provinents d'Apache Kafka.
    

## Notes importants:

- L'activitat s'ha de realitzar en **grups de 2 membres**. 
  Assegura't de saber qui és el teu company abans de començar l'activitat.
- Tot i que és possible completar les activitats directament en aquest quadern, **desaconsellamos fer-ho** a causa de possibles problemes de rendiment del servidor. Veuràs que cada activitat està continguda en la seva pròpia cel·la, cosa que et permetrà copiar-la fàcilment a un fitxer Python. Aquest fitxer pot ser **executat al servidor utilitzant el terminal de Jupyterlab, SSH o VSCode**. Un cop hagis executat i provat l'script amb èxit, simplement copia'l de nou a la cel·la corresponent del quadern. Aquest enfocament garanteix una execució més fluida i una millor gestió dels recursos del servidor.
- En alguns exercicis, necessitaràs fer **captures de pantalla per justificar les teves respostes**. 
  Pots capturar imatges utilitzant les eines del sistema operatiu que estiguis usant: "Retalls" a         Windows, "Imprimir pantalla", `Ctrl+C` en seleccionar una imatge, etc. 
  Un cop capturades, pots enganxar les imatges directament a les cel·les de resposta usant `Ctrl+V` o el menú     contextual que apareix en fer clic dret, que permetrà enganxar la imatge del porta-retalls. 
  Per visualitzar la imatge, has d'executar la cel·la.
- **Has d'utilitzar únicament les biblioteques proporcionades, tret que s'indiqui el contrari.**
- Si us plau, no canviïs el nom del quadern ni el tipus de les cel·les.

# Part I: Introducció a Apache Kafka (1.2 punts)

[Kafka](https://kafka.apache.org) és una plataforma distribuïda per gestionar 
esdeveniments en streaming que ens permet llegir, escriure i processar esdeveniments 
(registres o missatges, segons la terminologia de Kafka) distribuïts al llarg d'un clúster.

Començarem l'activitat creant un tema anomenat `activity2<usuari>`
al servei de Kafka del nostre clúster (substitueix `<usuari>` pel teu nom d'usuari). 
Un tema és una col·lecció ordenada d'esdeveniments que s'emmagatzema de forma persistent, generalment en disc,
i es distribueix i replica. Kafka tracta cada tema en cada partició com un registre 
(un conjunt ordenat de missatges). Cada missatge en una partició té un desplaçament únic, 
i aquests missatges tenen un període de retenció predeterminat de 7 dies (604.800.000 ms), 
tot i que pots modificar-lo en el moment de la creació del tema.

El broker de Kafka és `eimtcld3node1`, accessible al port estàndard 9092.

### Exercici 1: Crear un tema amb Kafka (0.15 punts)

Crea un tema de Kafka anomenat `activity2<usuari>` al nostre clúster amb un 
factor de replicació d'1 i una única partició, cosa que significa que usarem 
un sol node per emmagatzemar els missatges que rep Kafka. 
A més, especifica que els missatges només s'emmagatzemin durant 3 hores al tema.

In [1]:
!whoami

nlopezlopez2


In [2]:
!kafka-topics.sh --create --topic activity2nlopezlopez2 --bootstrap-server eimtcld3node1:9092 --replication-factor 1 --partitions 1 --config retention.ms=10800000

Error while executing topic command : Topic 'activity2nlopezlopez2' already exists.
[2026-05-02 20:44:46,877] ERROR org.apache.kafka.common.errors.TopicExistsException: Topic 'activity2nlopezlopez2' already exists.
 (kafka.admin.TopicCommand$)


### Exercici 2: Llista els temes de Kafka (0.15 punts)

Consulta el *tema* que acabes de crear i mostra'l.

In [3]:
!kafka-topics.sh --list --bootstrap-server eimtcld3node1:9092

Actividad_Batch
BDM_MASTODON
DD_mastodon_toots_counts
DD_mastodon_toots_lang
Ej11_1_mpadrellp_carlosibasam
Ej11_2_mpadrellp_carlosibasam
GettingSchema
Mastodon
MastodonTootCount
NOM_DEL_TOPIC
POSA_AQUI_EL_TOPIC_DE_MASTODON
TOPIC
__consumer_offsets
actividad2mmendezmartin0
actividad2monisor
activity2
activity2-ddonoso000
activity2-mmartinezmartin5
activity2Mikel
activity2TU_USUARIO
activity2_rcacace
activity2aalbama
activity2adelafuenteort
activity2adominguezriv
activity2adrianyfer
activity2aegeat
activity2aespinrearte
activity2aestanymac
activity2agarciagarcia17
activity2agarridovill
activity2agilif
activity2ahernandezmai
activity2alaraac
activity2alvaroams
activity2amedinau
activity2amedinay
activity2amontesinosn
activity2apidevall
activity2arojaslop
activity2clopezrojas0
activity2cmercerc
activity2coros
activity2cvilabu
activity2dcanete
activity2descalas
activity2eduardgc
activity2etovarn
activity2fcardonago
activity2fernandezmarc1997
activity2fferretar
activity2fherediama
activity2g

### Exercici 3: Esborra el tema de Kafka (0.15 punts)

Esborra el *tema* que vas crear a l'Exercici 1

In [4]:
!kafka-topics.sh --delete --topic activity2nlopezlopez2 --bootstrap-server eimtcld3node1:9092

### Exercici 4: Descriu el tema de Kafka (0.15 punts)

Crea el *tema* de nou tal com vas fer a l'Exercici 1 i utilitza la línia de comandes de kafka per descriure'l.

In [5]:
# Creem el tema de Kafka amb una partició, replicació 1 i retenció de 3 hores
!kafka-topics.sh --create --topic activity2nlopezlopez2 --bootstrap-server eimtcld3node1:9092 --replication-factor 1 --partitions 1 --config retention.ms=10800000

Created topic activity2nlopezlopez2.


In [6]:
# Mostrem la informació del tema creat
!kafka-topics.sh --describe --topic activity2nlopezlopez2 --bootstrap-server eimtcld3node1:9092

Topic: activity2nlopezlopez2	TopicId: rRaIqNQ5T6-81WPh8sRtAA	PartitionCount: 1	ReplicationFactor: 1	Configs: compression.type=producer,min.insync.replicas=1,segment.bytes=1073741824,retention.ms=10800000,max.message.bytes=1000000,index.interval.bytes=4096,retention.bytes=-1,segment.index.bytes=10485760
	Topic: activity2nlopezlopez2	Partition: 0	Leader: 1001	Replicas: 1001	Isr: 1001


In [ ]:
#He creat de nou el tema activity2nlopezlopez2 amb una única partició i un 
#factor de replicació de 1. Després, he utilitzat la comanda describe per 
#consultar la informació del tema i he pogut veure les seves característiques, 
#com el nombre de particions, el factor de replicació i la configuració de 
#retenció dels missatges (10800000 ms).

### Exercici 5: Crea un productor a Kafka (0.3 punts)

Crearem un esdeveniment al tema. Recorda que aquesta comanda <b>s'ha d'executar
des del terminal per interactuar</b>. Recorda usar `CTRL+c` quan hagis acabat d'enviar els missatges.

In [ ]:
!kafka-console-producer.sh --bootstrap-server eimtcld3node1:9092 --topic activity2nlopezlopez2

>

In [7]:
# En aquest exercici he utilitzat el productor de Kafka des del terminal
# amb la comanda kafka-console-producer.sh per enviar missatges al topic
# activity2nlopezlopez2. Un cop executada la comanda, apareix el símbol ">"
# que indica que el productor està esperant dades, i he anat escrivint
# diferents línies de text, que es publiquen com a missatges dins del topic.
# Finalment, he aturat el productor utilitzant CTRL+C.

### Exercici 6: Crea un consumidor a Kafka (0.3 punts)

Finalment, se't demana consultar els missatges enviats a través del terminal 
utilitzant el consumidor incorporat de Kafka per al tema. Específicament, has d'executar 
un consumidor connectant-te als diferents brokers existents i especificant el tema 
i la partició als quals s'han enviat els missatges. Pots obrir dos terminals 
i verificar que els missatges enviats amb el productor al broker poden ser 
visualitzats amb el consumidor de consola.

In [ ]:
!kafka-console-consumer.sh --bootstrap-server eimtcld3node1:9092 --topic activity2nlopezlopez2 --partition 0 --from-beginning


# En aquest exercici he utilitzat el consumidor de Kafka des del terminal
# amb la comanda kafka-console-consumer.sh per llegir els missatges del topic
# activity2nlopezlopez2. He indicat el broker i la partició corresponent,
# i amb l'opció --from-beginning he pogut veure tots els missatges enviats
# anteriorment pel productor. També he comprovat que, si envio nous missatges,
# aquests apareixen en temps real al consumidor.

# Part II: Ingesta de dades amb Apache Kafka (1 punt)

Per automatitzar la generació i el consum de dades, és comú treballar 
amb un llenguatge de programació com Python, en lloc de fer-ho directament en Bash. 
En les preguntes següents, explorarem la funcionalitat de Kafka utilitzant Python amb la biblioteca per defecte, que **NO HAS D'INSTAL·LAR, ja que ja està disponible en la versió correcta**. 
Tota la documentació associada amb l'API proporcionada la pots trobar a 
[Kafka](https://kafka-python.readthedocs.io/en/master/). Començarem amb el que 
ja hem cobert: escriure en el tema de Kafka.
Per a això, configurarem el [Kafka
producer](https://kafka-python.readthedocs.io/en/master/apidoc/KafkaProducer.html)
que inserirà valors numèrics en un tema de Kafka cada 3 segons. 
Mentre el productor escriu, procedirem a llegir els missatges a la pregunta 8.

### Exercici 7: Escriu un tema a Kafka (0.3 punts)


Se't demana escriure una seqüència numèrica de 250 nombres (de l'1 al 250) 
en el tema de Kafka activity2<usuari> que acabem de crear. Cada un dels 
missatges escrits en el tema ha de contenir informació sobre el tema on s'escriuen, 
una clau i el valor binari del nombre a escriure (per exemple, value=b'287'). 
És essencial revisar l'API associada amb el 
[Kafka producer](https://kafka-python.readthedocs.io/en/master/apidoc/KafkaProducer.html) 
per completar aquests exercicis.

Per guiar-te en l'exercici, et proporcionem una plantilla que pots usar per completar-lo. Has de completar les parts que falten del codi a la cel·la de codi a continuació.
   

In [ ]:
from kafka import KafkaProducer
import numpy as np
<FILL_IN>
for i in range(1,250):
    <FILL_IN>
producer.flush()

In [5]:
!docker ps

/bin/bash: docker: no s'ha trobat l'ordre


In [1]:
from kafka import KafkaProducer

producer = KafkaProducer(
    bootstrap_servers='eimtcld3node1:9092'
)

topic = 'activity2nlopezlopez2'

for i in range(1, 251):
    producer.send(
        topic,
        key=str(i).encode('utf-8'),
        value=str(i).encode('utf-8')
    )

producer.flush()
producer.close()

print("els missatjes s'han enviat correctament")

els missatjes s'han enviat correctament


### Exercici 8: Llegir un tema de Kafka (0.7 punts)

Usant la biblioteca de Python per a Kafka, 
[Kafka](https://kafka-python.readthedocs.io/en/master/),
llegeix els missatges enviats a l'exercici 7, mostrant només els valors, 
no les altres propietats del missatge. És important revisar l'ús de Kafka en 
([Python](https://kafka-python.readthedocs.io/en/master/usage.html))
i els paràmetres del consumidor de 
[Kafka
consumer](https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html).

Com en l'exercici anterior, aquí et proporcionem una plantilla que pots usar per completar l'exercici. Has de completar les parts que falten del codi.

In [ ]:
from kafka import KafkaConsumer
<FILL_IN>
for message in consumer:
    <FILL_IN>

In [2]:
from kafka import KafkaConsumer

consumer = KafkaConsumer(
    'activity2nlopezlopez2',
    bootstrap_servers='eimtcld3node1:9092',
    auto_offset_reset='earliest',
    group_id=None,
    consumer_timeout_ms=5000
)

for message in consumer:
    print(message.value.decode('utf-8'))

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250


##### Part III: Processament de dades en temps real amb Apache Spark Streaming (6.5 punts)

En aquesta part de l'activitat, ens centrarem en l'ús d'Apache Spark per processar dades en temps real. Aquí, utilitzarem:

-   [Spark Streaming](https://downloads.apache.org/spark/docs/3.3.0/streaming-programming-guide.html):
    Spark Streaming és un motor de processament de fluxos escalable i tolerant a fallades, 
    construït sobre Apache Spark. Permet el processament de dades en temps real amb alta capacitat de               rendiment i baixa latència. Amb Spark Streaming, pots realitzar anàlisis en temps real, 
    aprenentatge automàtic i processament de grafs sobre dades en streaming.
-   [Spark Structured Streaming](https://downloads.apache.org/spark/docs/3.3.0/structured-streaming-programming-guide.html):
    Structured Streaming és una API d'alt nivell per al processament de fluxos en Spark. Proporciona una             interfície declarativa i similar a SQL per processar dades d'streaming estructurades.
    Amb Structured Streaming, pots escriure consultes d'streaming que s'integren de manera 
    fluida amb el processament per lots, cosa que et permet construir pipelines de processament 
    de dades d'extrem a extrem.
   

En aquesta part de l'activitat, ens centrarem en la xarxa social Mastodon per processar les seves dades en streaming. És una bona idea familiaritzar-se amb l'estructura JSON d'un "toot" (consulta l'enllaç [Mastodon\'s API
webpage](https://docs.joinmastodon.org/entities/Status/)), per facilitar la comprensió dels diferents exercicis que proposem.

**Notes importants:**

-   Tingues en compte que, en usar la versió 3.3.0 d'Apache Spark, pot haver-hi diferències en l'API i altres característiques en comparació amb les versions més recents. Assegura't de consultar la documentació específica per a aquesta versió quan necessitis més detalls.

-   Recorda adjuntar una captura de pantalla del resultat de l'execució de cada exercici en aquesta part de l'activitat.

### Exercici 10: Spark Streaming (2.7 punts)

En aquest exercici, analitzaràs l'activitat a Mastodon comptant els toots en una finestra de temps. Comptar és una de les operacions fonamentals a Spark, i en aquesta activitat, aprofitarem la biblioteca de Spark Streaming per millorar les nostres capacitats de recompte.

#### Exercici 10.1: Counting in Windows (0.4 punts)

Com sabràs, la biblioteca Spark Streaming **processa les dades utilitzant el concepte de time windows**, 
agrupant els elements de dades segons el temps en què van ser rebuts. Aquest enfocament 
permet el processament per lots de dades d'streaming, habilitant anàlisis **sobre intervals de temps diferents**. Veuràs que la sintaxi per realitzar operacions sobre els RDDs dins d'aquestes finestres de temps és **pràcticament equivalent a les operacions estàndard de RDD** que ja coneixes.

Completa el codi a continuació per obtenir **el nombre de toots originals publicats cada 10 segons**. Exclou els retweets del teu recompte. Pot ser que necessitis consultar [Mastodon
API](https://docs.joinmastodon.org/entities/Status/) per entendre com estan estructurats els toots.

In [ ]:
import findspark
findspark.init()

import json
from pyspark import SparkContext
from pyspark.streaming import StreamingContext

# Initialize SparkContext and StreamingContext with a 1-second batch interval
app_name = "Activity2_Ex10_1"

# Create the SparkContext
try:
    sc = SparkContext("local[*]", appName=app_name)
except ValueError:
    sc.stop()
    sc = SparkContext("local[*]", appName=app_name)

sc.setLogLevel("ERROR")

# Create the StreamingContext
batch_interval = 5
ssc = StreamingContext(sc, batch_interval)
ssc.checkpoint("checkpoint")

# Define stream parameters
socket_host = "localhost"
socket_port = 9999

kafkaStream = ssc.socketTextStream(socket_host, socket_port)

# Count original toots every 10 seconds, excluding retoots/boosts
tootCounts = kafkaStream \
    .map(lambda x: json.loads(x.strip())) \
    .filter(lambda toot: toot.get("reblog") is None) \
    .window(10, 10) \
    .count()

# Print the count
tootCounts.pprint()

# Start the computation
try:
    ssc.start()
    ssc.awaitTermination()
except KeyboardInterrupt:
    ssc.stop()
    sc.stop()

SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/spark/jars/log4j-slf4j-impl-2.17.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/hadoop/lib/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/tez/lib/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]


:: loading settings :: url = jar:file:/usr/bigtop/3.3.0/usr/lib/spark/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml


Ivy Default Cache set to: /home/nlopezlopez2/.ivy2/cache
The jars for the packages stored in: /home/nlopezlopez2/.ivy2/jars
org.apache.spark#spark-sql-kafka-0-10_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-663268d4-ea52-4435-9332-6f5562c46eca;1.0
	confs: [default]
	found org.apache.spark#spark-sql-kafka-0-10_2.12;3.3.0 in central
	found org.apache.spark#spark-token-provider-kafka-0-10_2.12;3.3.0 in central
	found org.apache.kafka#kafka-clients;2.8.1 in central
	found org.lz4#lz4-java;1.8.0 in central
	found org.xerial.snappy#snappy-java;1.1.8.4 in central
	found org.slf4j#slf4j-api;1.7.32 in central
	found org.apache.hadoop#hadoop-client-runtime;3.3.2 in central
	found org.spark-project.spark#unused;1.0.0 in central
	found org.apache.hadoop#hadoop-client-api;3.3.2 in central
	found commons-logging#commons-logging;1.1.3 in central
	found com.google.code.findbugs#jsr305;3.0.0 in central
	found org.apache.commons#commons-pool2;2.11.1 in cen

-------------------------------------------
Time: 2026-05-02 21:07:30
-------------------------------------------
40



-------------------------------------------
Time: 2026-05-02 21:07:40
-------------------------------------------
40



-------------------------------------------
Time: 2026-05-02 21:07:50
-------------------------------------------
30



-------------------------------------------
Time: 2026-05-02 21:08:00
-------------------------------------------
36



-------------------------------------------
Time: 2026-05-02 21:08:10
-------------------------------------------
70



-------------------------------------------
Time: 2026-05-02 21:08:20
-------------------------------------------
42



-------------------------------------------
Time: 2026-05-02 21:08:30
-------------------------------------------
50



-------------------------------------------
Time: 2026-05-02 21:08:40
-------------------------------------------
39



-------------------------------------------
Time: 2026-05-02 21:08:50
-------------------------------------------
38



-------------------------------------------
Time: 2026-05-02 21:09:00
-------------------------------------------
39



-------------------------------------------
Time: 2026-05-02 21:09:10
-------------------------------------------
58



-------------------------------------------
Time: 2026-05-02 21:09:20
-------------------------------------------
45



-------------------------------------------
Time: 2026-05-02 21:09:30
-------------------------------------------
35



-------------------------------------------
Time: 2026-05-02 21:09:40
-------------------------------------------
40



-------------------------------------------
Time: 2026-05-02 21:09:50
-------------------------------------------
32



-------------------------------------------
Time: 2026-05-02 21:10:00
-------------------------------------------
26



-------------------------------------------
Time: 2026-05-02 21:10:10
-------------------------------------------
59



-------------------------------------------
Time: 2026-05-02 21:10:20
-------------------------------------------
50



-------------------------------------------
Time: 2026-05-02 21:10:30
-------------------------------------------
41



-------------------------------------------
Time: 2026-05-02 21:10:40
-------------------------------------------
18



-------------------------------------------
Time: 2026-05-02 21:10:50
-------------------------------------------
32



-------------------------------------------
Time: 2026-05-02 21:11:00
-------------------------------------------
51



-------------------------------------------
Time: 2026-05-02 21:11:10
-------------------------------------------
53



-------------------------------------------
Time: 2026-05-02 21:11:20
-------------------------------------------
54



-------------------------------------------
Time: 2026-05-02 21:11:30
-------------------------------------------
49



-------------------------------------------
Time: 2026-05-02 21:11:40
-------------------------------------------
38



-------------------------------------------
Time: 2026-05-02 21:11:50
-------------------------------------------
43



-------------------------------------------
Time: 2026-05-02 21:12:00
-------------------------------------------
63



-------------------------------------------
Time: 2026-05-02 21:12:10
-------------------------------------------
41



-------------------------------------------
Time: 2026-05-02 21:12:20
-------------------------------------------
46



-------------------------------------------
Time: 2026-05-02 21:12:30
-------------------------------------------
45



-------------------------------------------
Time: 2026-05-02 21:12:40
-------------------------------------------
54



-------------------------------------------
Time: 2026-05-02 21:12:50
-------------------------------------------
33



-------------------------------------------
Time: 2026-05-02 21:13:00
-------------------------------------------
42



-------------------------------------------
Time: 2026-05-02 21:13:10
-------------------------------------------
41



[Stage 0:>                                                          (0 + 1) / 1]

-------------------------------------------
Time: 2026-05-02 21:13:20
-------------------------------------------
38



-------------------------------------------
Time: 2026-05-02 21:13:30
-------------------------------------------
68



-------------------------------------------
Time: 2026-05-02 21:13:40
-------------------------------------------
48



-------------------------------------------
Time: 2026-05-02 21:13:50
-------------------------------------------
25



-------------------------------------------
Time: 2026-05-02 21:14:00
-------------------------------------------
32



-------------------------------------------
Time: 2026-05-02 21:14:10
-------------------------------------------
57



-------------------------------------------
Time: 2026-05-02 21:14:20
-------------------------------------------
43



-------------------------------------------
Time: 2026-05-02 21:14:30
-------------------------------------------
36



-------------------------------------------
Time: 2026-05-02 21:14:40
-------------------------------------------
28



-------------------------------------------
Time: 2026-05-02 21:14:50
-------------------------------------------
39



-------------------------------------------
Time: 2026-05-02 21:15:00
-------------------------------------------
43



-------------------------------------------
Time: 2026-05-02 21:15:10
-------------------------------------------
62



-------------------------------------------
Time: 2026-05-02 21:15:20
-------------------------------------------
50



-------------------------------------------
Time: 2026-05-02 21:15:30
-------------------------------------------
63



-------------------------------------------
Time: 2026-05-02 21:15:40
-------------------------------------------
53



-------------------------------------------
Time: 2026-05-02 21:15:50
-------------------------------------------
41



-------------------------------------------
Time: 2026-05-02 21:16:00
-------------------------------------------
47



-------------------------------------------
Time: 2026-05-02 21:16:10
-------------------------------------------
54



-------------------------------------------
Time: 2026-05-02 21:16:20
-------------------------------------------
50



-------------------------------------------
Time: 2026-05-02 21:16:30
-------------------------------------------
54



-------------------------------------------
Time: 2026-05-02 21:16:40
-------------------------------------------
28



-------------------------------------------
Time: 2026-05-02 21:16:50
-------------------------------------------
44



-------------------------------------------
Time: 2026-05-02 21:17:00
-------------------------------------------
42



-------------------------------------------
Time: 2026-05-02 21:17:10
-------------------------------------------
46



-------------------------------------------
Time: 2026-05-02 21:17:20
-------------------------------------------
50



-------------------------------------------
Time: 2026-05-02 21:17:30
-------------------------------------------
64



-------------------------------------------
Time: 2026-05-02 21:17:40
-------------------------------------------
33



-------------------------------------------
Time: 2026-05-02 21:17:50
-------------------------------------------
43



-------------------------------------------
Time: 2026-05-02 21:18:00
-------------------------------------------
37



-------------------------------------------
Time: 2026-05-02 21:18:10
-------------------------------------------
42



-------------------------------------------
Time: 2026-05-02 21:18:20
-------------------------------------------
48



-------------------------------------------
Time: 2026-05-02 21:18:30
-------------------------------------------
45



-------------------------------------------
Time: 2026-05-02 21:18:40
-------------------------------------------
41



-------------------------------------------
Time: 2026-05-02 21:18:50
-------------------------------------------
42



Afegeix la captura de pantalla del resultat **aquí**
-------------------------------------------
Time: 2026-05-02 21:10:50
-------------------------------------------
32

                                                                                
-------------------------------------------
Time: 2026-05-02 21:11:00
-------------------------------------------
51

                                                                                
-------------------------------------------
Time: 2026-05-02 21:11:10
-------------------------------------------
53

                                                                                
-------------------------------------------
Time: 2026-05-02 21:11:20
-------------------------------------------
54

Afegeix la captura de pantalla del resultat **aquí**

#### Exercici 10.2: Comptant Toots per Idioma (0.4 punts)

Com has observat a l'Exercici 2.1, el procés és bastant similar a treballar amb RDDs. 
Ara, aprofundirem en una anàlisi més complexa **comptant quants toots originals es creen 
per idioma cada 10 segons**. Per millorar la llegibilitat, et demanem que ordenïs 
els idiomes en ordre descendent segons el nombre de toots i limitïs la sortida als 5 principals idiomes.

In [ ]:
import findspark
findspark.init()

from pyspark import SparkContext
from pyspark.streaming import StreamingContext
import json

app_name = "Activity2_Ex10_2"

try:
    sc = SparkContext("local[*]", appName=app_name)
except ValueError:
    sc.stop()
    sc = SparkContext("local[*]", appName=app_name)

sc.setLogLevel("ERROR")

batch_interval = 5
ssc = StreamingContext(sc, batch_interval)
ssc.checkpoint("checkpoint")

socket_host = "localhost"
socket_port = 9999

kafkaStream = ssc.socketTextStream(socket_host, socket_port)

tootLangCounts = kafkaStream \
    .map(lambda x: json.loads(x.strip())) \
    .filter(lambda toot: toot.get("reblog") is None) \
    .map(lambda toot: (toot.get("language", "unknown"), 1)) \
    .window(10, 10) \
    .reduceByKey(lambda a, b: a + b) \
    .transform(lambda rdd: rdd.sortBy(lambda x: x[1], ascending=False)
                        .zipWithIndex()
                        .filter(lambda x: x[1] < 5)
                        .map(lambda x: x[0]))

tootLangCounts.pprint(5)

try:
    ssc.start()
    ssc.awaitTermination()
except KeyboardInterrupt:
    ssc.stop()
    sc.stop()

SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/spark/jars/log4j-slf4j-impl-2.17.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/hadoop/lib/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/tez/lib/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]


:: loading settings :: url = jar:file:/usr/bigtop/3.3.0/usr/lib/spark/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml


Ivy Default Cache set to: /home/nlopezlopez2/.ivy2/cache
The jars for the packages stored in: /home/nlopezlopez2/.ivy2/jars
org.apache.spark#spark-sql-kafka-0-10_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-f0994c87-d520-431f-baf1-1ba0852aefda;1.0
	confs: [default]
	found org.apache.spark#spark-sql-kafka-0-10_2.12;3.3.0 in central
	found org.apache.spark#spark-token-provider-kafka-0-10_2.12;3.3.0 in central
	found org.apache.kafka#kafka-clients;2.8.1 in central
	found org.lz4#lz4-java;1.8.0 in central
	found org.xerial.snappy#snappy-java;1.1.8.4 in central
	found org.slf4j#slf4j-api;1.7.32 in central
	found org.apache.hadoop#hadoop-client-runtime;3.3.2 in central
	found org.spark-project.spark#unused;1.0.0 in central
	found org.apache.hadoop#hadoop-client-api;3.3.2 in central
	found commons-logging#commons-logging;1.1.3 in central
	found com.google.code.findbugs#jsr305;3.0.0 in central
	found org.apache.commons#commons-pool2;2.11.1 in cen

-------------------------------------------
Time: 2026-05-02 21:35:15
-------------------------------------------
('en', 19)
(None, 4)
('es', 3)
('ja', 2)
('nl', 2)



-------------------------------------------
Time: 2026-05-02 21:35:25
-------------------------------------------
('en', 24)
(None, 8)
('fr', 5)
('de-de', 4)
('ja', 2)



-------------------------------------------
Time: 2026-05-02 21:35:35
-------------------------------------------
('en', 23)
(None, 5)
('ja', 3)
('de-de', 2)
('de', 2)



-------------------------------------------
Time: 2026-05-02 21:35:45
-------------------------------------------
('en', 21)
('es', 4)
('ja', 3)
('de', 3)
(None, 2)



-------------------------------------------
Time: 2026-05-02 21:35:55
-------------------------------------------
('en', 26)
(None, 5)
('de', 4)
('es', 3)
('nl', 3)



-------------------------------------------
Time: 2026-05-02 21:36:05
-------------------------------------------
('en', 33)
(None, 13)
('de', 3)
('es', 2)
('nl', 2)



-------------------------------------------
Time: 2026-05-02 21:36:15
-------------------------------------------
('en', 29)
('ja', 4)
('fr', 4)
(None, 2)
('es', 2)



-------------------------------------------
Time: 2026-05-02 21:36:25
-------------------------------------------
('en', 27)
(None, 8)
('fr', 6)
('de', 5)
('es', 2)



-------------------------------------------
Time: 2026-05-02 21:36:35
-------------------------------------------
('en', 19)
(None, 7)
('de', 5)
('fr', 4)
('ar', 1)



-------------------------------------------
Time: 2026-05-02 21:36:45
-------------------------------------------
('en', 24)
(None, 5)
('de', 3)
('es', 2)
('ja', 2)



-------------------------------------------
Time: 2026-05-02 21:36:55
-------------------------------------------
('en', 18)
(None, 4)
('ja', 3)
('de', 2)
('fr', 2)



-------------------------------------------
Time: 2026-05-02 21:37:05
-------------------------------------------
('en', 22)
(None, 6)
('ru', 2)
('es', 1)
('pl', 1)



-------------------------------------------
Time: 2026-05-02 21:37:15
-------------------------------------------
('en', 27)
('ja', 6)
(None, 4)
('de', 4)
('es', 3)



-------------------------------------------
Time: 2026-05-02 21:37:25
-------------------------------------------
('en', 18)
(None, 10)
('ja', 5)
('es', 3)
('fr', 3)



-------------------------------------------
Time: 2026-05-02 21:37:35
-------------------------------------------
('en', 20)
(None, 10)
('fr', 2)
('es', 1)
('ja', 1)



-------------------------------------------
Time: 2026-05-02 21:37:45
-------------------------------------------
('en', 14)
(None, 4)
('ja', 2)
('de', 2)
('pt', 2)



-------------------------------------------
Time: 2026-05-02 21:37:55
-------------------------------------------
('en', 20)
(None, 7)
('ja', 3)
('sv', 2)
('de', 1)



-------------------------------------------
Time: 2026-05-02 21:38:05
-------------------------------------------
('en', 19)
('de', 4)
('ja', 3)
(None, 1)
('es', 1)



-------------------------------------------
Time: 2026-05-02 21:38:15
-------------------------------------------
('en', 28)
(None, 5)
('ja', 4)
('de', 4)
('es', 2)



-------------------------------------------
Time: 2026-05-02 21:38:25
-------------------------------------------
('en', 24)
(None, 7)
('de', 3)
('nl', 3)
('es', 1)



-------------------------------------------
Time: 2026-05-02 21:38:35
-------------------------------------------
('en', 27)
('de', 8)
('es', 5)
(None, 3)
('tr', 1)



-------------------------------------------
Time: 2026-05-02 21:38:45
-------------------------------------------
('en', 18)
(None, 17)
('de', 5)
('fr', 3)
('ja', 2)



-------------------------------------------
Time: 2026-05-02 21:38:55
-------------------------------------------
('en', 21)
(None, 11)
('nl', 2)
('es', 1)
('ar', 1)



-------------------------------------------
Time: 2026-05-02 21:39:05
-------------------------------------------
('en', 26)
(None, 4)
('es', 1)
('ru', 1)
('de', 1)



-------------------------------------------
Time: 2026-05-02 21:39:15
-------------------------------------------
('en', 27)
(None, 4)
('ja', 1)
('sl', 1)
('fo', 1)



-------------------------------------------
Time: 2026-05-02 21:39:25
-------------------------------------------
('en', 31)
(None, 7)
('de', 5)
('es', 4)
('nl', 4)



-------------------------------------------
Time: 2026-05-02 21:39:35
-------------------------------------------
('en', 17)
(None, 11)
('de', 4)
('ja', 2)
('cs', 1)



-------------------------------------------
Time: 2026-05-02 21:39:45
-------------------------------------------
('en', 19)
(None, 12)
('de', 4)
('fr', 3)
('ja', 2)



-------------------------------------------
Time: 2026-05-02 21:39:55
-------------------------------------------
('en', 20)
('de', 4)
('nl', 3)
('fr', 3)
('es', 2)



-------------------------------------------
Time: 2026-05-02 21:40:05
-------------------------------------------
('en', 20)
(None, 9)
('de', 4)
('ru', 3)
('es', 3)



-------------------------------------------
Time: 2026-05-02 21:40:15
-------------------------------------------
('en', 38)
(None, 5)
('de', 4)
('pt', 3)
('ru', 2)



-------------------------------------------
Time: 2026-05-02 21:40:25
-------------------------------------------
('en', 31)
(None, 5)
('ja', 5)
('pt', 3)
('ru', 1)



-------------------------------------------
Time: 2026-05-02 21:40:35
-------------------------------------------
('en', 29)
(None, 6)
('de', 6)
('ar', 2)
('pt', 2)



-------------------------------------------
Time: 2026-05-02 21:40:45
-------------------------------------------
('en', 40)
(None, 5)
('de', 4)
('fr', 2)
('ru', 1)



-------------------------------------------
Time: 2026-05-02 21:40:55
-------------------------------------------
('en', 26)
('fr', 4)
(None, 3)
('nl', 3)
('de', 2)



-------------------------------------------
Time: 2026-05-02 21:41:05
-------------------------------------------
('en', 19)
(None, 6)
('de', 4)
('ar', 2)
('ja', 2)



-------------------------------------------
Time: 2026-05-02 21:41:15
-------------------------------------------
('en', 30)
(None, 8)
('es', 4)
('de', 2)
('fr', 2)



[Stage 0:>                                                          (0 + 1) / 1]

Afegeix la captura de pantalla del resultat **aquí** -------------------------------------------
Time: 2026-05-02 21:38:45
-------------------------------------------
('en', 18)
(None, 17)
('de', 5)
('fr', 3)
('ja', 2)

                                                                                
-------------------------------------------
Time: 2026-05-02 21:38:55
-------------------------------------------
('en', 21)
(None, 11)
('nl', 2)
('es', 1)
('ar', 1)

                                                                                
-------------------------------------------
Time: 2026-05-02 21:39:05
-------------------------------------------
('en', 26)
(None, 4)
('es', 1)
('ru', 1)
('de', 1)

#### Exercici 10.3: Mantenint el Recompte (0.4 punts)


Fins ara, hem estat obtenint resultats específics de lots, cosa que generalment és útil. 
Però, què passa si vols obtenir una visió més àmplia, mantenint la informació a través de les 
finestres per, per exemple, acumular tendències al llarg del temps? 
Aquest és l'enfocament de la nostra propera exploració.

En aquest exercici, t'invitem a modificar l'script anterior per **mantenir un recompte acumulat de tots els toots originals, categoritzats per idioma**. En lloc de simplement comptar nous toots cada cinc segons, anem a **sumar-los contínuament**. Pensa-hi com un marcador que s'actualitza constantment amb el nombre total de toots originals en cada idioma des del moment en què vam començar a fer streaming.

Per aconseguir-ho, treballarem amb les transformacions 
[**stateful transformations in Spark Streaming**](https://downloads.apache.org/spark/docs/3.3.0/streaming-programming-guide.html#caching--persistence).
Aquesta és una forma sofisticada de dir que recordarem les dades passades i les utilitzarem 
en els nostres càlculs actuals. És similar a mantenir un total acumulat en una variable 
global en lloc de començar des de zero cada vegada.

**T'invitem a completar el següent script:**

In [ ]:
import findspark
findspark.init()

from pyspark import SparkContext
from pyspark.streaming import StreamingContext
import json

app_name = "Activity2_Ex10_3"

try:
    sc = SparkContext("local[*]", appName=app_name)
except ValueError:
    sc.stop()
    sc = SparkContext("local[*]", appName=app_name)

sc.setLogLevel("ERROR")

batch_interval = 5
ssc = StreamingContext(sc, batch_interval)
ssc.checkpoint("checkpoint_10_3")

socket_host = "localhost"
socket_port = 9999

kafkaStream = ssc.socketTextStream(socket_host, socket_port)

def updateFunction(newValues, runningCount):
    if runningCount is None:
        runningCount = 0
    return sum(newValues) + runningCount

tootCounts = kafkaStream \
    .map(lambda x: json.loads(x.strip())) \
    .filter(lambda toot: toot.get("reblog") is None) \
    .map(lambda toot: (toot.get("language", "unknown"), 1)) \
    .updateStateByKey(updateFunction) \
    .transform(lambda rdd: rdd.sortBy(lambda x: x[1], ascending=False))

tootCounts.pprint()

try:
    ssc.start()
    ssc.awaitTermination()
except KeyboardInterrupt:
    ssc.stop()
    sc.stop()

SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/spark/jars/log4j-slf4j-impl-2.17.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/hadoop/lib/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/bigtop/3.3.0/usr/lib/tez/lib/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.apache.logging.slf4j.Log4jLoggerFactory]


:: loading settings :: url = jar:file:/usr/bigtop/3.3.0/usr/lib/spark/jars/ivy-2.5.1.jar!/org/apache/ivy/core/settings/ivysettings.xml


Ivy Default Cache set to: /home/nlopezlopez2/.ivy2/cache
The jars for the packages stored in: /home/nlopezlopez2/.ivy2/jars
org.apache.spark#spark-sql-kafka-0-10_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-3edbaace-d250-4a88-8c62-15cb121c7c71;1.0
	confs: [default]
	found org.apache.spark#spark-sql-kafka-0-10_2.12;3.3.0 in central
	found org.apache.spark#spark-token-provider-kafka-0-10_2.12;3.3.0 in central
	found org.apache.kafka#kafka-clients;2.8.1 in central
	found org.lz4#lz4-java;1.8.0 in central
	found org.xerial.snappy#snappy-java;1.1.8.4 in central
	found org.slf4j#slf4j-api;1.7.32 in central
	found org.apache.hadoop#hadoop-client-runtime;3.3.2 in central
	found org.spark-project.spark#unused;1.0.0 in central
	found org.apache.hadoop#hadoop-client-api;3.3.2 in central
	found commons-logging#commons-logging;1.1.3 in central
	found com.google.code.findbugs#jsr305;3.0.0 in central
	found org.apache.commons#commons-pool2;2.11.1 in cen

-------------------------------------------
Time: 2026-05-02 23:12:50
-------------------------------------------



-------------------------------------------
Time: 2026-05-02 23:12:55
-------------------------------------------
('en', 5)
('ja', 1)



-------------------------------------------
Time: 2026-05-02 23:13:00
-------------------------------------------
('en', 11)
('es', 1)
('ja', 1)
('pt', 1)
('fr', 1)



-------------------------------------------
Time: 2026-05-02 23:13:05
-------------------------------------------
('en', 20)
(None, 3)
('ja', 3)
('pt', 2)
('fr', 2)
('es', 1)
('ru', 1)



-------------------------------------------
Time: 2026-05-02 23:13:10
-------------------------------------------
('en', 33)
(None, 7)
('ja', 3)
('fr', 3)
('pt', 2)
('es', 1)
('ru', 1)
('ar', 1)



-------------------------------------------
Time: 2026-05-02 23:13:15
-------------------------------------------
('en', 48)
(None, 9)
('fr', 6)
('ja', 3)
('pt', 3)
('es', 1)
('ru', 1)
('ar', 1)
('de', 1)



-------------------------------------------
Time: 2026-05-02 23:13:20
-------------------------------------------
('en', 61)
(None, 14)
('fr', 6)
('ja', 5)
('de', 3)
('pt', 3)
('es', 2)
('ru', 1)
('ar', 1)



-------------------------------------------
Time: 2026-05-02 23:13:25
-------------------------------------------
('en', 69)
(None, 17)
('fr', 6)
('ja', 5)
('es', 4)
('de', 4)
('pt', 4)
('ru', 1)
('ar', 1)
('nl', 1)
...



-------------------------------------------
Time: 2026-05-02 23:13:30
-------------------------------------------
('en', 78)
(None, 17)
('fr', 7)
('ja', 6)
('es', 5)
('de', 5)
('pt', 4)
('ru', 1)
('ar', 1)
('gl', 1)
...



-------------------------------------------
Time: 2026-05-02 23:13:35
-------------------------------------------
('en', 96)
(None, 23)
('fr', 9)
('ja', 8)
('es', 6)
('de', 6)
('pt', 4)
('sk', 2)
('it', 2)
('ru', 1)
...



-------------------------------------------
Time: 2026-05-02 23:13:40
-------------------------------------------
('en', 107)
(None, 26)
('ja', 12)
('de', 9)
('fr', 9)
('es', 7)
('pt', 5)
('ar', 2)
('sk', 2)
('it', 2)
...



-------------------------------------------
Time: 2026-05-02 23:13:45
-------------------------------------------
('en', 115)
(None, 28)
('ja', 12)
('fr', 11)
('de', 10)
('es', 9)
('pt', 5)
('it', 5)
('fi', 2)
('ar', 2)
...



-------------------------------------------
Time: 2026-05-02 23:13:50
-------------------------------------------
('en', 121)
(None, 32)
('ja', 12)
('de', 12)
('fr', 11)
('es', 10)
('pt', 5)
('it', 5)
('fi', 2)
('ar', 2)
...



-------------------------------------------
Time: 2026-05-02 23:13:55
-------------------------------------------
('en', 132)
(None, 34)
('ja', 12)
('de', 12)
('fr', 11)
('es', 10)
('pt', 5)
('it', 5)
('sv', 3)
('fi', 2)
...



-------------------------------------------
Time: 2026-05-02 23:14:00
-------------------------------------------
('en', 136)
(None, 36)
('ja', 13)
('de', 13)
('fr', 11)
('es', 10)
('pt', 5)
('it', 5)
('sv', 3)
('fi', 2)
...



-------------------------------------------
Time: 2026-05-02 23:14:05
-------------------------------------------
('en', 155)
(None, 37)
('ja', 16)
('de', 15)
('es', 11)
('fr', 11)
('pt', 6)
('it', 6)
('sv', 3)
('sk', 3)
...



-------------------------------------------
Time: 2026-05-02 23:14:10
-------------------------------------------
('en', 175)
(None, 40)
('de', 18)
('ja', 16)
('es', 12)
('fr', 11)
('it', 7)
('pt', 6)
('sv', 3)
('sk', 3)
...



-------------------------------------------
Time: 2026-05-02 23:14:15
-------------------------------------------
('en', 191)
(None, 44)
('de', 20)
('ja', 17)
('es', 12)
('fr', 11)
('it', 7)
('pt', 6)
('tr', 3)
('sv', 3)
...



Afegeix la captura de pantalla del resultat **aquí** -------------------------------------------
Time: 2026-05-02 23:14:10
-------------------------------------------
('en', 175)
(None, 40)
('de', 18)
('ja', 16)
('es', 12)
('fr', 11)
('it', 7)
('pt', 6)
('sv', 3)
('sk', 3)
...

                                                                                
-------------------------------------------
Time: 2026-05-02 23:14:15
-------------------------------------------
('en', 191)
(None, 44)
('de', 20)
('ja', 17)
('es', 12)
('fr', 11)
('it', 7)
('pt', 6)
('tr', 3)
('sv', 3)
...

#### Exercici 10.4: Windowed Counting (0.5 punts)

Com has observat, Spark Streaming és increïblement flexible i 
fàcil d'usar, i aquí tens un truc interessant que pot realitzar: 
**et permet trobar un punt intermedi entre comptar toots en una [time
window](https://downloads.apache.org/spark/docs/3.3.0/streaming-programming-guide.html#window-operations)
i mantenir un total acumulat**. Imaginem que volem crear un panell de control, 
com un tauler, que mostri el nombre de toots realitzats en cada idioma. 
El gir és que **volem que aquesta actualització es produeixi cada 5 segons, 
però estem rastrejant els recomptes durant un minut complet**.

Així, cada 5 segons, el nostre panell s'actualitza, proporcionant-nos 
el darrer recompte acumulat durant un minut. És com tenir un marcador en viu 
que s'actualitza amb freqüència i també realitza un seguiment del que ha ocorregut 
en els darrers 60 segons, no només en els darrers 5. D'aquesta manera, obtens 
tant actualitzacions immediates com una vista més àmplia del que està ocorrent, 
tot al mateix temps. Mostra només els 10 principals idiomes.

**Modifica el següent script per aconseguir aquest objectiu:**


In [ ]:
import findspark
findspark.init()

from pyspark import SparkContext
from pyspark.streaming import StreamingContext
import json

app_name = "Activity2_Ex10_3"

try:
    sc = SparkContext("local[*]", appName=app_name)
except ValueError:
    sc.stop()
    sc = SparkContext("local[*]", appName=app_name)

sc.setLogLevel("ERROR")

batch_interval = 5
ssc = StreamingContext(sc, batch_interval)
ssc.checkpoint("checkpoint_10_3")

socket_host = "localhost"
socket_port = 9999

kafkaStream = ssc.socketTextStream(socket_host, socket_port)

def updateFunction(newValues, runningCount):
    if runningCount is None:
        runningCount = 0
    return sum(newValues) + runningCount

tootCounts = kafkaStream \
    .map(lambda x: json.loads(x.strip())) \
    .filter(lambda toot: toot.get("reblog") is None) \
    .map(lambda toot: (toot.get("language", "unknown"), 1)) \
    .updateStateByKey(updateFunction) \
    .transform(lambda rdd: rdd.sortBy(lambda x: x[1], ascending=False))

tootCounts.pprint()

try:
    ssc.start()
    ssc.awaitTermination()
except KeyboardInterrupt:
    ssc.stop()
    sc.stop()

Afegeix la captura de pantalla del resultat **aquí**

### Exercici 10.5: Powering Up (1 punt)


Molt bé, ja sabem que els RDDs en Spark són increïblement versàtils: 
són com la navalla suïssa del processament de dades d'stream. Pots fer pràcticament 
qualsevol operació amb ells. Tanmateix, **a mesura que les coses es tornen més complexes, 
el repte augmenta.**

Ara, anem a fer el nostre panell de control encara més interessant. En lloc de només mostrar 
quants toots tenim per minut, afegim algunes característiques noves. 
No seria interessant **rastrejar la longitud mitjana d'aquests toots?**
I hi ha més: anem a descobrir **qui és l'usuari amb més seguidors entre els que han tootejat en aquell minut.**

Espera, **¡hi ha més!** Per fer tota aquesta informació súper fàcil de llegir i entendre, la presentarem en un format de taula ordenat i net. No es tracta només de les dades, sinó de fer-les amigables per a l'usuari i visualment digeribles.

La taula resultant s'ha d'actualitzar en intervals de 5 segons, i les finestres de mitjana han de ser de 60 segons. Les columnes d'aquesta taula han de ser:
    
-   **`lang`:** Idioma
-   **`num_toots`:** Nombre de toots originals en aquell idioma
-   **`avg_len_content`:** Longitud mitjana (en caràcters) del contingut del toot
-   **`user`:** Usuari amb més seguidors
-   **`followers`:** Nombre de seguidors d'aquest usuari

Per fer la sortida més llegible, limita el nombre de files a 10.

**TIP:** Hi ha un exemple molt útil a [Spark Streaming](https://downloads.apache.org/spark/docs/3.3.0/streaming-programming-guide.html#dataframe-and-sql-operations). Busca'l!


In [ ]:
<FILL IN>

Afegeix la captura de pantalla del resultat **aquí**

## Exercici 11: Structured Streaming (2.6 punts)

Com has vist en l'últim exercici, depenent de les operacions, 
l'API de Spark Streaming pot no resultar tan convenient, 
especialment perquè has de tractar amb APIs de baix nivell. Afortunadament, **Spark provides a
high-level API called [Spark Structured
Streaming](https://downloads.apache.org/spark/docs/3.3.0/structured-streaming-programming-guide.html)** que et permet **expressar els càlculs d'streaming de la mateixa manera que expressaries un càlcul per lots sobre dades estructurades estàtiques**, com les que usaries en el processament per lots.

En aquest conjunt d'exercicis, et submergirás en el fascinant món de Spark Structured Streaming.

## Exercici 11.1: Getting the Schema (0.3 punts)

Una de les característiques més interessants de Spark Structured Streaming és 
**com gestiona les dades estructurades**. Per exemple, el flux de dades en el nostre tema de Kafka, 
on **cada "toot" arriba en un format JSON ordenat**, cosa que el fa estructurat i organitzat.

De manera similar a com es treballa amb DataFrames a Spark, **Structured Streaming usa esquemes de dades per analitzar dades estructurades**, essencialment un plànol de com estan disposades les dades. Per al processament per lots, Spark pot sovint deduir aquesta estructura directament de les dades. Tanmateix, amb les dades d'streaming, és una mica diferent: **hem de definir aquesta estructura prèviament.**

En els exercicis següents, utilitzarem un truc convenient: **en lloc de definir manualment**
l'estructura complexa d'un "toot", inicialment **extraurem alguns toots de Kafka i els analitzarem en un lot per aprendre el seu esquema**. És com fer un cop d'ull per entendre com estan organitzades les coses. 
Un cop tinguem l'esquema, l'aplicarem a la nostra pipeline d'streaming.

La teva tasca en aquest exercici és realitzar aquesta transformació. Després, utilitzant les operacions de DataFrame amb les quals ja estàs familiaritzat, crearàs una taula amb les columnes següents que ens permetrà veure els toots individualment a mesura que es processen:


-   **`id`:** Identificador únic per a cada toot
-   **`created_at`:** Data i hora en què es va crear el toot
-   **`content`:** El contingut del toot
-   **`language`:** L'idioma del toot
-   **`username`:** El nom d'usuari de l'autor del toot
-   **`followers_count`:** Nombre de seguidors de l'autor.


Recorda que ens **interessen els toots originals**. Filtra els que són retweets.

Un altre aspecte fonamental que has de gestionar aquí és **seleccionar el outputMode adequat**. Consulta 
[la documentació](https://downloads.apache.org/spark/docs/3.3.0/structured-streaming-programming-guide.html#output-modes)
i tria el que millor s'adapti a aquest exercici.


In [ ]:
import findspark
findspark.init()

from pyspark.sql import SparkSession
from pyspark.sql.types import StringType, StructType, StructField
from pyspark import SparkConf, SparkContext
from pyspark.sql.functions import json_tuple, from_json, col
from pyspark.sql import SparkSession

conf = SparkConf()
conf.setMaster("local[*]")
sc = SparkContext(conf=conf)
sc.setLogLevel("ERROR")

# Initialize Spark Session for Structured Streaming
app_name = "activity3_1" + <FILLIN> # Replace with your Spark app name must include the username of the members of the group

spark = SparkSession \
    .builder \
    .appName(app_name) \
    .getOrCreate()

# Define Kafka parameters
kafka_topic = <FILLIN>
kafka_bootstrap_servers = <FILLIN>  # Replace with your Kafka bootstrap servers

# Read a small batch of data from Kafka for schema inference!
batch_df = spark \
    .read \
    .format("kafka") \
    .option("kafka.bootstrap.servers", kafka_bootstrap_servers) \
    .option("subscribe", kafka_topic) \
    .option("startingOffsets", "earliest") \
    .option("endingOffsets", "{\"" + kafka_topic + "\":{\"0\":10}}") \
    .load()

# Infer schema
schema = spark.read.json(batch_df.selectExpr("CAST(value AS STRING)").rdd.map(lambda x: x[0])).schema

# Create streaming DataFrame by reading data from Kafka
toots = spark \
    .readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", kafka_bootstrap_servers) \
    .option("subscribe", kafka_topic) \
    .option("startingOffsets", <FILLIN>) \
    .load()

# Parse the value column as JSON and apply the infered schema. Then select the columns we need.
toots_df = toots\
    .select(from_json(col("value").cast("string"), schema).alias("parsed_value"))\
    .select(<FILLIN>)

try:
    # Open stream to console (you need to execute it in a terminal to see the output)
    query = toots_df \
            .writeStream \
            .outputMode(<FILLIN>) \
            .format("console")\
            .start()

    query.awaitTermination()
except KeyboardInterrupt:
    query.stop()
    spark.stop()
    sc.stop()

Afegeix la captura de pantalla del resultat **aquí**

### Exercici 11.2: Agregant Dades des d'un Flux (0.3 punts)

Spark Structured Streaming és realment poderós, especialment quan es realitzen operacions sobre un flux continu de dades. En aquest exercici, aprofundirem en Structured Spark Streaming, centrant-nos específicament en l'agregació de dades des d'un flux de Kafka. És similar al que vam fer a l'Exercici 2. La teva missió és **comptar el nombre de toots originals en cada idioma.**

Aquí et mostrem com ha de ser la teva sortida:

-   **`language`:** Aquesta columna mostra l'idioma dels toots.
-   **`count`:** Aquí és on mostraràs el nombre de toots per cada idioma.

La teva taula ha d'**acumular aquests recomptes cada 10 segons, i també ha de continuar acumulant-los**. A més, fes-la amigable per a l'usuari **ordenant els idiomes pel nombre de toots, amb els idiomes més parladors a la part superior.**

Ara, aquí hi ha una **part clau** d'aquest exercici: **has de triar el mode de sortida adequat per a la teva consulta d'streaming**. Recorda, el **mode de sortida determina com s'escriu cada lot de dades resultant al destí de sortida**. Les teves opcions són els modes 'Complete', 'Append' i 'Update'. Pensa quin encaixa millor per al nostre escenari de recompte acumulatiu i ordenat. I **no oblidis escriure el teu raonament als comentaris.**


In [ ]:
import findspark
findspark.init()

from pyspark.sql import SparkSession
from pyspark.sql.types import StringType, StructType, StructField
from pyspark import SparkConf, SparkContext
from pyspark.sql.functions import from_json, col, window
from pyspark.sql import SparkSession

conf = SparkConf()
conf.setMaster("local[*]")
sc = SparkContext(conf=conf)
sc.setLogLevel("ERROR")

# Initialize Spark Session for Structured Streaming
app_name = "activity3_2" + <FILLIN> # Replace with your Spark app name must include the username of the members of the group

spark = SparkSession \
    .builder \
    .appName(app_name) \
    .getOrCreate()

# Define Kafka parameters
kafka_topic = <FILLIN>
kafka_bootstrap_servers = <FILLIN>  # Replace with your Kafka bootstrap servers

# Read a small batch of data from Kafka for schema inference!
batch_df = spark \
    .read \
    .format("kafka") \
    .option("kafka.bootstrap.servers", kafka_bootstrap_servers) \
    .option("subscribe", kafka_topic) \
    .option("startingOffsets", "earliest") \
    .option("endingOffsets", "{\"" + kafka_topic + "\":{\"0\":10}}") \
    .load()

# Infer schema
schema = spark.read.json(batch_df.selectExpr("CAST(value AS STRING)").rdd.map(lambda x: x[0])).schema

# Create streaming DataFrame by reading data from Kafka
toots = spark \
    .readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", kafka_bootstrap_servers) \
    .option("subscribe", kafka_topic) \
    .option("startingOffsets", <FILLIN>) \
    .load()

# Parse the value column as JSON and apply the infered schema. Then select the columns we need.
toots_df = toots\
    .select(from_json(col("value").cast("string"), schema).alias("parsed_value"))\
    .filter(<FILLIN>)\
    .select(<FILLIN>)\
    .groupBy(<FILLIN>)\
    .<FILLIN>
    ...
    <FILLIN>


try:
    # Open stream to console (you need to execute it in a terminal to see the output)
    query = toots_df \
            .writeStream \
            .outputMode(<FILLIN>) \
            .format("console")\
            .<FILLIN>
            ...
            .<FILLIN>
            .trigger(<FILLIN>)\
            .start()\

    query.awaitTermination()
except KeyboardInterrupt:
    query.stop()
    spark.stop()
    sc.stop()

Afegeix la captura de pantalla del resultat **aquí**

### Exercici 11.3: Windowed Counting (1 punt)

Bon treball! Has après a realitzar agregacions i a fer un seguiment dels recomptes al llarg del temps. Com vas notar a l'Exercici 2.4, de vegades és més efectiu mantenir aquests recomptes en **finestres de temps específiques**. Ara, volem que apliquis aquesta tècnica utilitzant [functions available in Spark Structured
Streaming](https://downloads.apache.org/spark/docs/3.3.0/structured-streaming-programming-guide.html#window-operations-on-event-time). 
Tingues en compte que Spark Structured Streaming gestiona el temps de manera diferent a Spark Streaming, per la qual cosa hauràs de considerar-ho en analitzar i interpretar els resultats.

La teva tasca és **crear una taula que mostri un recompte de la quantitat de toots originals (recorda filtrar els retweets) realitzats en cada idioma, segmentats dins d'un marc de temps específic.** Per a aquest exercici, has d'usar una sliding window d'un minut, amb les dades refrescant-se cada 5 segons. Aquest enfocament et permetrà monitorar de prop la freqüència dels toots en diferents idiomes al llarg d'intervals breus i superposats.

Et demanem que proporcionis una taula amb l'estructura següent:

- **`window`:** Mostra el rang de temps
- **`language`:** Aquesta columna mostra l'idioma dels toots.
- **`count`:** Aquí és on mostraràs el nombre de toots per a cada idioma.

Els resultats s'han d'**ordenar per finestra de temps i recompte en ordre descendent.**


In [ ]:
import findspark
findspark.init()

from pyspark.sql import SparkSession
from pyspark.sql.types import StringType, StructType, StructField
from pyspark import SparkConf, SparkContext
from pyspark.sql.functions import json_tuple, from_json, col, window
from pyspark.sql import SparkSession

conf = SparkConf()
conf.setMaster("local[*]")
sc = SparkContext(conf=conf)
sc.setLogLevel("ERROR")

# Initialize Spark Session for Structured Streaming
app_name = "activity3_3" + <FILLIN> # Replace with your Spark app name must include the username of the members of the group

spark = SparkSession \
    .builder \
    .appName(app_name) \
    .getOrCreate()

# Define Kafka parameters
kafka_topic = <FILLIN>
kafka_bootstrap_servers = <FILLIN>  # Replace with your Kafka bootstrap servers

# Read a small batch of data from Kafka for schema inference!
batch_df = spark \
    .read \
    .format("kafka") \
    .option("kafka.bootstrap.servers", kafka_bootstrap_servers) \
    .option("subscribe", kafka_topic) \
    .option("startingOffsets", "earliest") \
    .option("endingOffsets", "{\"" + kafka_topic + "\":{\"0\":10}}") \
    .load()

# Infer schema
schema = spark.read.json(batch_df.selectExpr("CAST(value AS STRING)").rdd.map(lambda x: x[0])).schema

# Create streaming DataFrame by reading data from Kafka
toots = spark \
    .readStream \
    .format("kafka") \
    .option("kafka.bootstrap.servers", kafka_bootstrap_servers) \
    .option("subscribe", kafka_topic) \
    .option("startingOffsets", <FILLIN>) \
    .load()

# Parse the value column as JSON and apply the infered schema. Then select the columns we need.
toots_df = toots\
    .select(from_json(col("value").cast("string"), schema).alias("parsed_value"))\
    .filter(<FILLIN>)\
    .select(<FILLIN>)\
    .groupBy(
        window(<FILLIN>),
        <FILLIN>
    )\
    .count()\
    .orderBy(<FILLIN>, <FILLIN>, ascending=False)

try:
    # Open stream to console (you need to execute it in a terminal to see the output)
    query = toots_df \
            .writeStream \
            .outputMode(<FILLIN>) \
            .format("console")\
            .option("truncate", "false")\
            .trigger(<FILLIN>)\
            .start()\

    query.awaitTermination()
except KeyboardInterrupt:
    query.stop()
    spark.stop()
    sc.stop()

Afegeix la captura de pantalla del resultat **aquí**

### Exercici 11.4: Unir Fluxos (1 punt)

En aquest darrer exercici, explorarem una característica molt interessant de Spark Streaming que et permet 
**[unir dos streams](https://downloads.apache.org/spark/docs/3.3.0/structured-streaming-programming-guide.html#stream-stream-joins)
i analitzar-los!**


Per simplificar les coses, ja et proporcionem dos fluxos de dades pre-agregats. 
El primer, que es troba al tema de Kafka **`mastodon_toots_original_domain`**, mostra el recompte de toots originals per a diverses instàncies de Mastodon (recorda que Mastodon és una federació d'instàncies) durant **una finestra fixa d'un minut**. 
El segon flux, al **`mastodon_toots_retoot_domain` topic**, presenta dades similars, però per als toots que són retweets (compartits) d'altres toots. Les dades emmagatzemades als temes de Kafka tenen la mateixa estructura en format JSON:

- Una estructura `window` amb dos elements de tipus `string`: `start` i `end` 
- Un component `string` anomenat `mastodon_instance` amb el topic.
- Un element de tipus `integer` anomenat `count` amb el nombre de toots realitzats en aquell domini en el rang de temps específic.

Donat que l'estructura de les dades és bastant senzilla, aquesta vegada **et demanem que la defineixis manualment**, en lloc d'usar el truc de batch. Un cop hagis configurat les estructures, **hauràs d'obrir un flux per a cada font de Kafka.** El pas següent és unir aquests fluxos. Volem que realitzis una **left join del flux de toots originals amb el flux de retweets**. Després de completar la unió, la teva sortida ha d'incloure:


- **`window`:** el rang de temps
- **`mastodon_instance`:** el domini de la instància de Mastodon
- **`original_count`:** nombre de toots originals publicats en aquell domini durant aquell rang de temps
- **`retweet_count`:** nombre de toots de retweets publicats en aquell domini durant aquell rang de temps

**TIP**: Realitzar una unió en línia de dos fluxos no és una tasca fàcil, i hi ha moltes restriccions que has de respectar. 
Llegeix acuradament [la documentació](https://downloads.apache.org/spark/docs/3.3.0/structured-streaming-programming-guide.html#stream-stream-joins)
i recorda que estem usant la versió 2.4.0. 
A més, recorda que estem realitzant **join over time**, i aquest és un component **clau**. 
Conceptes com els que has après sobre windows són fonamentals juntament amb conceptes com [watermarking](https://downloads.apache.org/spark/docs/3.3.0/structured-streaming-programming-guide.html#stream-stream-joins).
I recorda, **els modes de sortida són complicats** i has de triar-ne un que sigui adequat per a la feina que vols fer.


In [ ]:
import findspark
findspark.init()

from pyspark.sql import SparkSession
from pyspark.sql.types import StringType, StructType, StructField, IntegerType
from pyspark import SparkConf, SparkContext
from pyspark.sql.functions import from_json, col, window, to_timestamp, struct
from pyspark.sql import SparkSession

# Initialize Spark Context
conf = SparkConf()
conf.setMaster("local[*]")
sc = SparkContext(conf=conf)
sc.setLogLevel("ERROR")

# Initialize Spark Session for Structured Streaming
app_name = "activity3_4" + <FILLIN> # Replace with your Spark app name must include the username of the members of the group

spark = SparkSession \
    .builder \
    .appName(app_name) \
    .getOrCreate()

# Define schema for the incoming data
schema = StructType(<FILLIN>)

# Define Kafka parameters
toots_original_topic = <FILLIN>
toots_retoot_topic = <FILLIN>
kafka_bootstrap_servers = <FILLIN>  # Replace with your Kafka bootstrap servers

# Create streaming DataFrame by reading original toots data from Kafka
toots_original = spark \
    .readStream \
    .<FILLIN>
    ...
    .<FILLIN>
    .load()

# Parse the value column as JSON and apply the infered schema. Then select the columns we need.
toots_original_df = toots_original\
    .<FILLIN>
    ...
    .<FILLIN>

# Create streaming DataFrame by reading retoots data from Kafka
toots_retoot = spark \
    .readStream \
    .<FILLIN>
    ...
    .<FILLIN>
    .load()

# Parse the value column as JSON and apply the infered schema. Then select the columns we need.
toots_retoot_df = toots_retoot\
    .<FILLIN>
    ...
    .<FILLIN>

# Join the two streams
toots_join_df = toots_original_df.join(<FILLIN>...<FILLIN>)

try:
    # Start running the query that prints the running counts to the console
    query = toots_join_df\
            .writeStream \
            <FILLIN>
            ...
            <FILLIN>
            .option("numRows", 100)\
            .start()\

    query.awaitTermination()
except KeyboardInterrupt:
    query.stop()
    spark.stop()
    sc.stop()

Afegeix la captura de pantalla del resultat **aquí**

### **Exercici 12**: Preguntes teòriques (*2.5 punts*)

Durant l'entrevista final amb cada grup, que es realitzarà mitjançant videotrucada, es revisarà breument la pràctica realitzada.

En l'última part de l'entrevista es plantejaran 4 preguntes teòriques relacionades amb els continguts de la pràctica i del mòdul 5.

- Cada pregunta tindrà una puntuació de 0,625 punts.

- Les preguntes es respondran oralment durant l'entrevista.

- Es valorarà la comprensió dels conceptes teòrics i la capacitat d'explicar-los amb claredat.

In [ ]:
# YOUR CODE HERE
raise NotImplementedError()